optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -1166,6 +1155,22 @@ public int countReinitializableVariables(ScoreDirector<Solution_> scoreDirector,
         return Iterators.concat(iteratorList.iterator());
     }
 
+    public Stream<Object> extractAllEntitiesStream(Solution_ solution) {
+        List<Stream<Object>> streamList = new ArrayList<>(
+                entityMemberAccessorMap.size() + entityCollectionMemberAccessorMap.size());
+        for (MemberAccessor memberAccessor : entityMemberAccessorMap.values()) {
+            Object entity = extractMemberObject(memberAccessor, solution);
+            if (entity != null) {
+                streamList.add(Stream.of(entity));
+            }
+        }
+        for (MemberAccessor memberAccessor : entityCollectionMemberAccessorMap.values()) {
+            Collection<Object> entityCollection = extractMemberCollectionOrArray(memberAccessor, solution, false);
+            streamList.add(entityCollection.stream());
+        }
+        return streamList.stream().flatMap(Function.identity());
```

Comment:
I think that you can do the same by using `Stream.concat()`. That will allow you to get rid of the `List`.

Reply:
It behaves like concat (= lazy), but concat() needs an array as input. We don't know how long the array will be in advance, so we need to create a list anyway :)

Reply:
The `List` isn't lazy either. :-) But you can replace `list.add(Stream.of(x))` by `stream = Stream.concat(stream, x)` and save yourself quite some code.

Reply:
ah, I didn't know, let me check that.

Reply:
yes, it works :)